feat: adds more summary information for kubernetes

cdrage opened the pull request:
feat: adds more summary information for kubernetes

### What does this PR do?

* Refactors and unifies some of the parts that were used for summary
  pages
* Adds a lot more information with regards to details on each Kubernetes
  artifact (ingress, routes, services, etc.)

### Screenshot / video of UI

<!-- If this PR is changing UI, please include
screenshots or screencasts showing the difference -->

https://github.com/containers/podman-desktop/assets/6422176/faf5f465-c829-43dc-8052-c8a8e5849108




### What issues does this PR fix or reference?

<!-- Include any related issues from Podman Desktop
repository (or from another issue tracker). -->

Closes https://github.com/containers/podman-desktop/issues/5658

### How to test this PR?

Click on the Summary for each section.

<!-- Please explain steps to reproduce -->

Signed-off-by: Charlie Drage <charlie@charliedrage.com>

cdrage commented:
This is up for review!

I do need to **write more tests** before this is merged.

But since this is a lot more information added, I'll wait until we get a final review before some simple tests are added.

cdrage commented:
> IMHO I feel it is hard to read, they are a lot of information, and not enough contrast in certain information.
> 
> 
> 
> **containers**
> 
> 
> 
> E.g. The containers in the pods are not contrasted enough, they are the core of the resource, and as a developers, I might want an indicator related to their status, (are they all running ? Did one is still pending?)
> 
> 
> 
> ![image](https://github.com/containers/podman-desktop/assets/42176370/4f80f1ef-9092-4f86-b8ce-a8bb8da5a7d4)
> 
> 
> 
> **IPs**
> 
> 
> 
> I also feel like _other tools_ are highlighting some informations, we might want to have some kind of `code block` for some information, and colors for status
> 
> 
> 
> ![image](https://github.com/containers/podman-desktop/assets/42176370/61d0f380-13e6-4e08-acd6-f3c76793024f)
> 
> 
> 
> 

I 100% and totally agree. This PR is meant to add the initial information. 

I am hoping in subsequent PRs we can work with the UI team to define some good ways to contrast all the new information once we have more resources.

benoitf commented on packages/renderer/src/lib/deployments/DeploymentDetailsSummary.svelte:
```diff
@@ -1,11 +1,10 @@
-import DeploymentColumnConditions from './DeploymentColumnConditions.svelte';
-import type { DeploymentUI } from './DeploymentUI';
-import DeploymentColumnPods from './DeploymentColumnPods.svelte';
+import ObjectMeta from '../kube/details/ObjectMeta.svelte';
```
I would prefix ObjectMeta with something like KubernetesDetailsObjectMeta to avoid to have like 5 ObjectMeta classes in Podman Desktop

benoitf commented on packages/renderer/src/lib/deployments/DeploymentDetailsSummary.svelte:
```diff
@@ -1,11 +1,10 @@
-import type { DeploymentUI } from './DeploymentUI';
-import DeploymentColumnPods from './DeploymentColumnPods.svelte';
+import ObjectMeta from '../kube/details/ObjectMeta.svelte';
+import Deployment from '../kube/details/Deployment.svelte';
```
please add prefix /suffix with Kubernetes and details so with the className you see in one sec if it's a generic, specific, etc class

benoitf commented on packages/renderer/src/lib/deployments/DeploymentDetailsSummary.svelte:
```diff
@@ -1,11 +1,10 @@
-import DeploymentColumnPods from './DeploymentColumnPods.svelte';
+import ObjectMeta from '../kube/details/ObjectMeta.svelte';
+import Deployment from '../kube/details/Deployment.svelte';
+import DeploymentStatus from '../kube/details/DeploymentStatus.svelte';
```
add kube/Kubernetes prefix

benoitf commented on packages/renderer/src/lib/ingresses-routes/IngressRouteDetailsSummary.svelte:
```diff
@@ -2,14 +2,18 @@
-import type { RouteUI } from './RouteUI';
-import IngressRouteColumnHostPath from './IngressRouteColumnHostPath.svelte';
-import IngressRouteColumnBackend from './IngressRouteColumnBackend.svelte';
+import ObjectMeta from '../kube/details/ObjectMeta.svelte';
+import Ingress from '../kube/details/Ingress.svelte';
+import IngressStatus from '../kube/details/IngressStatus.svelte';
```
need to find a new home for this objects

it looks strange to import from `../kube/details` when we're inside `ingresses-routes`

benoitf replied:
also we should favor the `/@/ notation`

benoitf commented on packages/renderer/src/lib/kube/details/ObjectMeta.svelte:
```diff
@@ -0,0 +1,122 @@
+
+export let artifact: V1ObjectMeta | undefined;
+
+if (artifact?.creationTimestamp) {
+  artifact.creationTimestamp = new Date(artifact.creationTimestamp);
+}
+
+let internalLabelsDropdownOpen: boolean = false;
+let internalAnnotationsDropdownOpen: boolean = false;
+
+// Sort labels out to internal and external
+let labels: [string, string][] = [];
+let internalLabels: [string, string][] = [];
+if (artifact?.labels) {
+  Object.entries(artifact.labels).forEach(([key, value]) => {
+    if (internalKubernetesKeys.includes(key)) {
+      internalLabels.push([key, value]);
+    } else {
+      labels.push([key, value]);
+    }
+  });
+}
+
+// Sort annotationsout to internal and external
+let annotations: [string, string][] = [];
+let internalAnnotations: [string, string][] = [];
+if (artifact?.annotations) {
+  Object.entries(artifact.annotations).forEach(([key, value]) => {
+    if (internalKubernetesKeys.includes(key)) {
+      internalAnnotations.push([key, value]);
+    } else {
+      annotations.push([key, value]);
+    }
+  });
+}
```
it feels weird that it's not inside onMount method ?

cdrage replied:
So because it is simple sorting, I did not put it in the onMount method since I get this error:

```
if_block0.p is not a function
```

When svelte tries to check: `  {#if annotations.length > 0}`

It seems to stem from this issue: https://github.com/sveltejs/svelte/issues/5789

benoitf commented on packages/renderer/src/lib/kube/details/Volume.svelte:
```diff
@@ -0,0 +1,227 @@
+<script lang="ts">
+/* There are many types of volumes that need to be shown, all with different 'name' values, etc.
```
this class needs a prefix, Volume is too generic

benoitf commented on packages/renderer/src/lib/kube/details/utils.ts:
```diff
@@ -0,0 +1,58 @@
+/**********************************************************************
+ * Copyright (C) 2022-2023 Red Hat, Inc.
```
2024

benoitf reviewed (commented):
nice improvement

I'm finding a bit strange the pattern
```
<tr>
    <td class="{rowStyle}">Name</td>
    <td class="{rowStyle}">{artifact.name}</td>
  </tr>
```

we're like duplicating <tr><td with the style 'rowStyle' everywhere

we have the table component maybe should use it

or we avoid all the <tr><td> stuff

I mean, you could have a <Row> or <WhateverComponent> object instead of duplicating all the lines

or just give an array like
```
 [["Metadata"], ['Name', artifact.name]]
 ```

 like a structure to be displayed

lstocchi commented on packages/renderer/src/lib/kube/details/KubeIngressArtifact.svelte:
```diff
@@ -0,0 +1,35 @@
+          {#if rule.http}
+            {#each rule.http.paths as path}
+              Path: {path.path} • Link:
+              <Link externalRef="{artifact.tls && artifact.tls.length > 0 ? 'https' : 'http'}://{rule.host}{path.path}">
+                {artifact.tls && artifact.tls.length > 0 ? 'https' : 'http'}://{rule.host}{path.path}
+              </Link>
+              • Backend: {path.backend.service?.name}{path.backend.service?.port?.number ? ':' : ''}{path.backend
+                .service?.port?.number}
```
Maybe we should check the values here.
I tried to check the summary of the resources i had on my cluster and i saw 
![image](https://github.com/containers/podman-desktop/assets/49404737/e7c67102-c9f0-4b86-8a7f-7faefc388cf4)

lstocchi reviewed (commented):
Nice and impressive work really. Everything is showed in a clear way.

Looking at the code i don't have much to say other than the thought Florent had about the repeating <tr>, if there could be a way to simplify the table stuff would be awesome.

Just noticed a missing check on the ingress using my existing cluster.

cdrage commented:
> Nice and impressive work really. Everything is showed in a clear way.
> 
> Looking at the code i don't have much to say other than the thought Florent had about the repeating , if there could be a way to simplify the table stuff would be awesome.
> 
> Just noticed a missing check on the ingress using my existing cluster.

Updated!

@lstocchi @benoitf @deboer-tim 

The reason for the "repeating" and the fact we don't implement a generic table is that we plan to implement some of the more "prettier" parts of the summary back. For example, how we show the columns for NodePort, etc:

![Screenshot 2024-02-14 at 8 41 44 AM](https://github.com/containers/podman-desktop/assets/6422176/cd5380c2-81e3-494c-89e9-6202af529432)

I've gone ahead and implemented the `Cell.svelte` to avoid repeating the styles however.

Another reason is that the API for Kubernetes is quite... finnicky.

There's empty arrays (on purpose), some elements are undefined and others are straight up empty strings, it's an absolute mess at times especially dealing with custom resources too.

So that's why we're trying to keep the "meta" stuff generic (using objectmeta, etc.), but the more larger svelte files such as Deployment, would have to be custom.

I hope that helps!

This PR is just the first stepping stone to improved summary pages.

cdrage commented:
Ready for another review 👍

odockal commented:
@cdrage Do we need some E2E tests to cover functionality or are unit tests enough?
Or maybe something in between?

cdrage commented:
> @cdrage Do we need some E2E tests to cover functionality or are unit tests enough? Or maybe something in between?

Unit tests are fine :) these are just taking the artifacts and parsing tghem.

lstocchi commented:
@cdrage thanks charlie, saw your update. 
The missing piece is still the backend from my cluster
![image](https://github.com/containers/podman-desktop/assets/49404737/14a76f3a-c762-4917-b44c-5661599dfa2c)

cdrage commented:
> @cdrage thanks charlie, saw your update. The missing piece is still the backend from my cluster ![image](https://private-user-images.githubusercontent.com/49404737/305094223-14a76f3a-c762-4917-b44c-5661599dfa2c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDgwMDg2NjMsIm5iZiI6MTcwODAwODM2MywicGF0aCI6Ii80OTQwNDczNy8zMDUwOTQyMjMtMTRhNzZmM2EtYzc2Mi00OTE3LWI0NGMtNTY2MTU5OWRmYTJjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjE1VDE0NDYwM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAwZmI0ZjNlZGViOGY2NWMzOGZkYjVjMjgwZWRhMTY1OWMyNjczNmJhNWI4NjZlMDllODAzNmJjMmMwOGJjNzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.QUwoDkd8u7WDy3eeEpcWc_lIttM84zdFaa5n-jCT_ao)

Odd you don't have a backend for the ingress! Or it may be a resource instead of a service?

I'm curious what your YAML is. 

I've gone ahead and updated the PR to also include showing resource as well as checking to see that 'service' and 'resource' is available before showing.

cdrage commented:
Since I've gotten a few reviews, I'm going to go ahead and add the rest of the tests now so it's ready for final review 👍

jeffmaury commented:
If got an empty details, then the section should be hidden ?

![image](https://github.com/containers/podman-desktop/assets/695993/df09f57f-8fb8-4fc7-afed-8ea30681aa8a)

cdrage commented:
> If got an empty details, then the section should be hidden ?
> 
> ![image](https://private-user-images.githubusercontent.com/695993/307011184-df09f57f-8fb8-4fc7-afed-8ea30681aa8a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg2MTA1NzksIm5iZiI6MTcwODYxMDI3OSwicGF0aCI6Ii82OTU5OTMvMzA3MDExMTg0LWRmMDlmNTdmLThmYjgtNGZjNy1hZmVkLThlYTMwNjgxYWE4YS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjIyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIyMlQxMzU3NTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wNTRmN2E0NThjNWJmYmRlMDJkMDNjM2M2NmQwMTU3ZDQ2NGZiYTUzMGZkNWY1MmIzZWM4OWExNTcxNGRlMDNkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.vUHkZs_FAXo20hzSCYVe8MvV2MaG-ZPQsKrB9Fj8Q5E)

Good catch!

Today I learned that you can have an ingress with zero rules and just a defaultBackend. 

I've updated the PR to accurately show the default resource (should appear as not blank for you anymore):
![Screenshot 2024-02-22 at 9 16 24 AM](https://github.com/containers/podman-desktop/assets/6422176/1ac77158-c25d-422c-b6fb-b6e59f2f4de1)

And this made me open up issue https://github.com/containers/podman-desktop/issues/6107

I tested the ingress (screenshot above) with this file:

```yaml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: test
  namespace: default
spec:
  defaultBackend:
    resource:
      name: 'foobar'
      kind: 'foobar'
```

feloy commented:
I find it unfortunate to hardcode the list of information displayed in these Summaries, in which order, with which titles, etc.

I would prefer to have a kind of template (remember gotemplates ;) ) and the resource would be applied to the template before to display it. This way, we can imagine in the future that users (or companies) may be able to provide their own templates.

But, anyway, that's a great first version

cdrage commented:
> I find it unfortunate to hardcode the list of information displayed in these Summaries, in which order, with which titles, etc.
> 
> I would prefer to have a kind of template (remember gotemplates ;) ) and the resource would be applied to the template before to display it. This way, we can imagine in the future that users (or companies) may be able to provide their own templates.
> 
> > There's empty arrays (on purpose), some elements are undefined and others are straight up empty strings, it's an absolute mess at times especially dealing with custom resources too.
> 
> For this, we could compute the value that would be displayed, but let the template decide where/if to display this data. That would help separate data and presentation.
> 
> But, anyway, that's a great work

Yeah :( I find it unfortunate too with regards to making it more generic / template-like. 

At least we do have some generic sections such as metadata.

We made an early decision to "manually" create the summary pages and not have them automated, as we will be eventually adding UI mockups such as:

![Screenshot 2024-02-22 at 9 54 18 AM](https://github.com/containers/podman-desktop/assets/6422176/777136e1-2ecf-4d0a-ac95-dd70256f05ca)

back into the summary pages. 

As well as the fact that the information is "curated" and not all information would be helpful to summarize / show to the user, such as managed fields, or other fields which are more kubernetes-internal-only.

If a user wanted the "raw" information, they can have a look at Inspect tab, but for Summary we try to make it as useful as possible / being able to look at all the information immediatley.

lstocchi reviewed (approved):
Tried again with many random resources and everything is displayed correctly and as expected for me. LGTM, great work!! 🚀

feloy commented on packages/renderer/src/lib/kube/details/KubeIngressArtifact.svelte:
```diff
@@ -0,0 +1,87 @@
+                </Link>
+              {:else}
+                <!-- If rule.host is undefined, we only show the path without constructing a link -->
+                Path only: {path.path}
```
Here, the path.path will be displayed twice (already on line 65)

feloy commented on packages/renderer/src/lib/kube/details/KubeVolumeArtifact.svelte:
```diff
@@ -0,0 +1,228 @@
+  {/if}
+  {#if artifact.emptyDir}
+    <tr>
+      <Cell>Empty Directory</Cell>
+      <Cell>Medium: {artifact.emptyDir.medium || 'Default'}</Cell>
```
The sizeLimit information would be helpful IMO

feloy reviewed (commented):
A couple of nits, LGTM otherwise

cdrage commented:
Thank you! 

I've added the two requests (path being shown twice and volumeSize) being added.

Reviewed by @lstocchi and @feloy 

I'm going to enable auto-merge and start to get this in as this will be the initial "base" implementation to improve on.

Any further improvements can be in new PR's related to https://github.com/containers/podman-desktop/pull/5828#issuecomment-1927117690 comment with regards to further improving it (showing IP, etc in code font, etc.).
